File: internal/analyzer/container.go
package analyzer

import (
	"context"
	"fmt"
	"log/slog"
	"sort"
	"strings"

	"infra-stats/internal/vmclient"
)

const containerGroupLabels = "envir, job, instance, name, department"

// RunContainers собирает аналитику по контейнерам (cadvisor) за все периоды.
// Возвращает все контейнеры с данными; решение «что показывать» — в notifier.
func (e *Engine) RunContainers(ctx context.Context) []ContainerStat {
	slog.Info("Starting container analysis",
		slog.Int("periods", len(e.periods)),
		slog.Any("periods", e.periods),
	)

	sel := containerSelector(e.containersCfg.Filters)

	limitCores, err := e.client.Query(ctx, fmt.Sprintf(
		`sum by (%s) ((container_spec_cpu_quota%s > 0) / (container_spec_cpu_period%s > 0))`,
		containerGroupLabels, sel, sel,
	))
	if err != nil {
		slog.Warn("Container CPU limit query failed", slog.String("error", err.Error()))
		limitCores = nil
	}
	limitByKey := indexContainerByKey(limitCores)

	nodeCores, err := e.client.Query(ctx, `count by (instance) (node_cpu_seconds_total{instance!=""})`)
	if err != nil {
		slog.Warn("Node CPU cores query failed", slog.String("error", err.Error()))
		nodeCores = nil
	}
	coresByInstance := indexByLabel(nodeCores, "instance")

	nodeMem, err := e.client.Query(ctx, `node_memory_MemTotal_bytes{instance!=""}`)
	if err != nil {
		slog.Warn("Node memory total query failed", slog.String("error", err.Error()))
		nodeMem = nil
	}
	memTotalByInstance := indexByLabel(nodeMem, "instance")

	stats := make(map[string]*ContainerStat)

	for _, period := range e.periods {
		usage, err := e.client.Query(ctx, fmt.Sprintf(
			`sum by (%s) (rate(container_cpu_usage_seconds_total%s[%s]))`,
			containerGroupLabels, sel, period,
		))
		if err != nil {
			slog.Warn("Container CPU usage query failed", slog.String("period", period), slog.String("error", err.Error()))
			continue
		}

		memRatio, err := e.client.Query(ctx, fmt.Sprintf(
			`avg_over_time((container_memory_working_set_bytes%s / container_spec_memory_limit_bytes%s)[%s:2m]) * 100`,
			sel, sel, period,
		))
		if err != nil {
			slog.Warn("Container memory ratio query failed", slog.String("period", period), slog.String("error", err.Error()))
			continue
		}

		wsAvg, err := e.client.Query(ctx, fmt.Sprintf(
			`avg_over_time(container_memory_working_set_bytes%s[%s:2m])`,
			sel, period,
		))
		if err != nil {
			slog.Warn("Container working set query failed", slog.String("period", period), slog.String("error", err.Error()))
			continue
		}

		for _, s := range usage {
			key := containerKeyOf(s.Metric)
			if !finite(s.Value) {
				continue
			}
			st := getOrCreateContainer(stats, s.Metric)

			if limitCores, ok := limitByKey[key]; ok && finite(limitCores) && limitCores > 0 {
				st.CPU = append(st.CPU, MetricValue{Period: period, Value: round(s.Value/limitCores*100, 1)})
			}
			if cores, ok := coresByInstance[s.Metric["instance"]]; ok && finite(cores) && cores > 0 {
				st.CPUVM = append(st.CPUVM, MetricValue{Period: period, Value: round(s.Value/cores*100, 1)})
			}
		}

		for _, s := range memRatio {
			if !finite(s.Value) {
				continue
			}
			st := getOrCreateContainer(stats, s.Metric)
			st.Memory = append(st.Memory, MetricValue{Period: period, Value: round(s.Value, 1)})
		}

		for _, s := range wsAvg {
			if !finite(s.Value) {
				continue
			}
			nodeTotal, ok := memTotalByInstance[s.Metric["instance"]]
			if !ok || !finite(nodeTotal) || nodeTotal <= 0 {
				continue
			}
			st := getOrCreateContainer(stats, s.Metric)
			st.MemVM = append(st.MemVM, MetricValue{Period: period, Value: round(s.Value/nodeTotal*100, 1)})
		}
	}

	containers := make([]ContainerStat, 0, len(stats))
	for _, st := range stats {
		containers = append(containers, *st)
	}
	sort.Slice(containers, func(i, j int) bool {
		if containers[i].Instance != containers[j].Instance {
			return containers[i].Instance < containers[j].Instance
		}
		return containers[i].Name < containers[j].Name
	})

	slog.Info("Container analysis complete", slog.Int("containers", len(containers)))
	return containers
}

func containerSelector(filters map[string]string) string {
	parts := []string{`name!=""`}
	keys := make([]string, 0, len(filters))
	for k := range filters {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	for _, k := range keys {
		parts = append(parts, fmt.Sprintf(`%s=%q`, k, filters[k]))
	}
	return "{" + strings.Join(parts, ",") + "}"
}

func containerKeyOf(m map[string]string) string {
	return strings.Join([]string{
		m["envir"], m["job"], m["instance"], m["name"], m["department"],
	}, "\x00")
}

func indexContainerByKey(series []vmclient.Series) map[string]float64 {
	out := make(map[string]float64, len(series))
	for _, s := range series {
		out[containerKeyOf(s.Metric)] = s.Value
	}
	return out
}

func indexByLabel(series []vmclient.Series, label string) map[string]float64 {
	out := make(map[string]float64, len(series))
	for _, s := range series {
		out[s.Metric[label]] = s.Value
	}
	return out
}

func getOrCreateContainer(stats map[string]*ContainerStat, m map[string]string) *ContainerStat {
	key := containerKeyOf(m)
	st, ok := stats[key]
	if !ok {
		st = &ContainerStat{
			Name:       strings.TrimPrefix(m["name"], "/"),
			Instance:   m["instance"],
			Job:        m["job"],
			Envir:      m["envir"],
			Department: m["department"],
		}
		stats[key] = st
	}
	return st
}

File: app/analyzer/container.py
"""Модуль анализа нагрузки контейнеров.

Формирует Prometheus-запросы для сбора метрик CPU, RAM и working-set
контейнеров и нод, собирает результаты в единый список ContainerStat
для каждого анализируемого периода.
"""
from __future__ import annotations

import asyncio
import logging
import math
from typing import Optional

from app.analyzer.analyzer import ContainerStat, Engine, MetricValue, instance_selector, round_value

logger = logging.getLogger("analyzer.container")


def container_selector(filters: dict[str, str], instances: Optional[list[str]] = None) -> str:
    """Собирает строку-селектор Prometheus для контейнеров.

    Объединяет переданные фильтры, исключает контейнеры с пустым
    именем и опционально фильтрует по списку инстансов.
    """
    parts = ['name!=""']
    for key in sorted(filters):
        parts.append(f'{key}="{filters[key]}"')
    parts.append(instance_selector(instances or []))
    return "{" + ",".join(parts) + "}"


def container_key_of(metric: dict[str, str], labels: Optional[list[str]] = None) -> str:
    """Возвращает уникальный ключ контейнера по набору меток.

    Ключ строится конкатенацией значений указанных лейблов через
    нулевой байт \\x00, что исключает коллизии.
    """
    if labels is None:
        labels = ["envir", "job", "instance", "name", "department"]
    return "\x00".join(metric.get(lbl, "") for lbl in labels)


def container_cpu_usage_query(sel: str, period: str, group_labels: str) -> str:
    """Формирует PromQL-запрос потребления CPU контейнерами.

    Вычисляет rate потребления секунд CPU за указанный период
    с группировкой по заданным лейблам.
    """
    return f"sum by ({group_labels}) (rate(container_cpu_usage_seconds_total{sel}[{period}]))"


def container_mem_ratio_query(sel: str, period: str) -> str:
    """Формирует PromQL-запрос отношения потребления RAM к лимиту.

    Использует средние значения за период без субзапроса,
    результат — процент от лимита памяти контейнера.
    """
    return (
        f"(avg_over_time(container_memory_working_set_bytes{sel}[{period}]) "
        f"/ avg_over_time(container_spec_memory_limit_bytes{sel}[{period}])) * 100"
    )


def container_ws_avg_query(sel: str, period: str) -> str:
    """Формирует PromQL-запрос среднего working-set памяти контейнера."""
    return f"avg_over_time(container_memory_working_set_bytes{sel}[{period}])"


def node_cores_query(inst_sel: str) -> str:
    """Формирует PromQL-запрос количества CPU-ядер ноды."""
    return f"count by (instance) (node_cpu_seconds_total{{{inst_sel}}})"


def node_mem_total_query(inst_sel: str) -> str:
    """Формирует PromQL-запрос общей оперативной памяти ноды."""
    return f"node_memory_MemTotal_bytes{{{inst_sel}}}"


def _index_container_by_key(series, labels: Optional[list[str]] = None) -> dict[str, float]:
    """Индексирует серию метрик по ключу контейнера."""
    return {container_key_of(s.metric, labels): s.value for s in series}


def _index_by_label(series, label: str) -> dict[str, float]:
    """Индексирует серию метрик по значению указанного лейбла."""
    return {s.metric.get(label): s.value for s in series}


def _get_or_create(
    stats: dict[str, ContainerStat], metric: dict[str, str], labels: Optional[list[str]] = None
) -> ContainerStat:
    """Возвращает существующий или создаёт новый ContainerStat по метке.

    Если для данного ключа контейнера запись уже есть — возвращает её,
    иначе создаёт новую запись с метаданными из metric и добавляет в словарь stats.
    """
    key = container_key_of(metric, labels)
    st = stats.get(key)
    if st is None:
        st = ContainerStat(
            name=metric.get("name", "").lstrip("/"),
            instance=metric.get("instance", ""),
            job=metric.get("job", ""),
            envir=metric.get("envir", ""),
            department=metric.get("department", ""),
        )
        stats[key] = st
    return st


async def run_containers(engine: Engine, instances: Optional[list[str]] = None) -> list[ContainerStat]:
    """Запускает полный анализ нагрузки контейнеров.

    Параллельно запрашивает лимиты CPU, количество ядер и объём RAM нод,
    затем для каждого анализируемого периода собирает метрики CPU, RAM
    и working-set. Возвращает отсортированный список ContainerStat
    с вычисленными значениями в процентах.
    """
    logger.info("Starting container analysis periods=%d periods=%s", len(engine.periods), engine.periods)

    group_labels = engine.containers_cfg.group_labels
    labels = [l.strip() for l in group_labels.split(",") if l.strip()]
    sel = container_selector(engine.containers_cfg.filters, instances)
    inst_sel = instance_selector(instances)

    limit_cores, node_cores, node_mem = await asyncio.gather(
        engine._q_cached(
            f"sum by ({group_labels}) ((container_spec_cpu_quota{sel} > 0) "
            f"/ (container_spec_cpu_period{sel} > 0))"
        ),
        engine._q_cached(node_cores_query(inst_sel)),
        engine._q_cached(node_mem_total_query(inst_sel)),
    )
    limit_by_key = _index_container_by_key(limit_cores or [], labels)
    cores_by_instance = _index_by_label(node_cores or [], "instance")
    mem_total_by_instance = _index_by_label(node_mem or [], "instance")

    stats: dict[str, ContainerStat] = {}

    for period in engine.periods:
        usage, mem_ratio, ws_avg = await asyncio.gather(
            engine._q(container_cpu_usage_query(sel, period, group_labels)),
            engine._q(container_mem_ratio_query(sel, period)),
            engine._q(container_ws_avg_query(sel, period)),
        )

        if usage is not None:
            for s in usage:
                if not math.isfinite(s.value):
                    continue
                st = _get_or_create(stats, s.metric, labels)

                limit = limit_by_key.get(container_key_of(s.metric, labels))
                if limit is not None and math.isfinite(limit) and limit > 0:
                    st.cpu.append(MetricValue(period=period, value=round_value(s.value / limit * 100, 1)))

                cores = cores_by_instance.get(s.metric.get("instance"))
                if cores is not None and math.isfinite(cores) and cores > 0:
                    st.cpu_vm.append(MetricValue(period=period, value=round_value(s.value / cores * 100, 1)))
        else:
            logger.warning("Container CPU usage query failed period=%s", period)

        if mem_ratio is not None:
            for s in mem_ratio:
                if not math.isfinite(s.value):
                    continue
                st = _get_or_create(stats, s.metric, labels)
                st.memory.append(MetricValue(period=period, value=round_value(s.value, 1)))
        else:
            logger.warning("Container memory ratio query failed period=%s", period)

        if ws_avg is not None:
            for s in ws_avg:
                if not math.isfinite(s.value):
                    continue
                node_total = mem_total_by_instance.get(s.metric.get("instance"))
                if node_total is None or not math.isfinite(node_total) or node_total <= 0:
                    continue
                st = _get_or_create(stats, s.metric, labels)
                st.mem_vm.append(MetricValue(period=period, value=round_value(s.value / node_total * 100, 1)))
        else:
            logger.warning("Container working set query failed period=%s", period)

    containers = sorted(stats.values(), key=lambda c: (c.instance, c.name))
    logger.info("Container analysis complete containers=%d", len(containers))
    return containers
